### PR TITLE
Fix plot_implicit for compound boolean regions

### LINKED ISSUE #17710: `plot_implicit` with `And` ignoring relations
With `sympy` 1.4 and Python 3.7.3, using `And` for intersecting relations in `plot_implicit` respects only one of the inequalities. Working example below.

```python
import sympy as sp
sp.var('x,y',real=True)
sp.plot_implicit((y > x)&(y > 2),y_var=y,x_var=x)
```

The resulting plot is below; it doesn't respect `y > 2`.

image

### LINKED ISSUE #17719: plot_implicit error for Xor
This can be an issue that needs to be resolved after addressing #17710 
`xor` is not properly comprehended for `plot_implicit` now, ~~but it can be beyond the scope for what I want to do in #17711~~

```
from sympy import *

x, y = symbols('x y')

a = (x - 1)**2 + y**2 < 2
b = (x + 1)**2 + y**2 < 2

plot_implicit(a ^ b, x, y)
```

### PR DESCRIPTION
**Pull request by sylee957**
#### References to other Issues or PRs

Fixes #17710
Fixes #17719

#### Brief description of what is fixed or changed

Python `and` or `or` operations between tuples does not operate in elementwise way
`(False, False) and (True, True) => (True, True)`
And I see the interval arithmetics module in the plotting module uses `tuple` for some fuzzy logic of an interval membership.
but in this way, it may not work properly with any compound boolean region specifications.

I have added a new object `intervalBool` instead, which can have the boolean operations properly defined.
Although I have quickly worked on this by just doing an naive elementwise applycation without much concern of the fuzzy values just to see if it works, but it can be improved in the further work.

Now, it works
```python
x, y = symbols('x y')
plot_implicit((y > 2) & (y > x), x_var=x, y_var=y)
```
image

#### Other comments

I also reverted some changes in #12108 because the real problem for bitwise boolean object may only happen only for the usage of intervals.

I may also change the naming like `intervalBool`. Maybe something like `intervalMembership` can be more better?

#### Release Notes

- plotting 
  - Fixed `plot_implicit` for composite boolean regions like `(y > 2) & (y > x)`
- printing
  - Fixed `str` to print `Xor` to the python operator `^`.

### CODE REVIEW

File: sympy/plotting/intervalmath/interval_arithmetic.py [hunk 1]
```diff
@@ -34,6 +34,79 @@
...
+
+        The second value is determined as follows:
+        - True: If both the intervals in comparison are valid.
+        - False: If at least one of the intervals is False, else
+        - None
+    """
+    def __init__(self, a, b):
+        self._wrapped = (a, b)
+
+    def __getitem__(self, i):
```
Comment (kenluck2001) [inline sympy/plotting/intervalmath/interval_arithmetic.py]: Add a check for one-off error if the index are beyond range

Comment (oscarbenjamin) [reply sympy/plotting/intervalmath/interval_arithmetic.py]: What needs to be checked for? It this is off by one then the IndexError below should handle it.

File: sympy/plotting/intervalmath/interval_arithmetic.py [hunk 2]
```diff
@@ -34,6 +34,84 @@
...
+        try:
+            return self._wrapped[i]
+        except IndexError:
+            raise IndexError(
```
Comment (oscarbenjamin) [inline sympy/plotting/intervalmath/interval_arithmetic.py]: I would rather let the original IndexError propagate upwards

File: sympy/plotting/intervalmath/interval_arithmetic.py (hunk 1 above)
Comment (sylee957) [reply sympy/plotting/intervalmath/interval_arithmetic.py]: I had the same thought, but I think that the drawback of making python index error to propagate upward can be the error message not being descriptive.
And I have seen this [thread](https://stackoverflow.com/questions/2522005/cost-of-exception-handlers-in-python) that the cost of using exception catching block for places where it is supposed to be programmed to have no errors is cheap.

**Comment by sylee957** [comment]
I may add some additional error tests for `intervalMembership` that are not handled in coverage,
and add some actual plotting tests.

However, since we don't check the actual outputs for the plots, I'm trying to use [image comparison](https://matplotlib.org/devel/testing.html#writing-an-image-comparison-test) if possible

File: sympy/plotting/intervalmath/interval_membership.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        return intervalMembership(fuzzy_and([a1, a2]), fuzzy_and([b1, b2]))
+
+    def __or__(self, other):
+        if not isinstance(other, intervalMembership):
+            raise ValueError(
+                "The comparison is not supported for {}.".format(other))
+
+        a1, b1 = self
+        a2, b2 = other
+        return intervalMembership(fuzzy_or([a1, a2]), fuzzy_or([b1, b2]))
```
Comment (sylee957) [inline sympy/plotting/intervalmath/interval_membership.py]: Whatever, I've probably got the logic wrong for the second value.
Maybe second value have to be always`fuzzy_and([b1, b2])`.

**Comment by sylee957** [comment]
I have written the image comparison tests using the crude `compare_images` because of the difficulty of implementing decorators in a compatible way, and also some file path errors that I have no idea to resolve.

I hope that it is really a public API like featured in the [doc](https://matplotlib.org/api/testing_api.html#matplotlib.testing.compare.compare_images), though I'm not clear about it since they have some other trivial docs exposed.

**Comment by oscarbenjamin** [comment]
Looks good!

**Pull request merged**